src/Features/Core/Portable/CodeRefactorings/AbstractRefactoringHelpersService.cs

```diff
@@ -0,0 +1,176 @@
+                do
+                {
+                    // consider either a Node that is a parent of touched Token (selection can be within) or ancestor Node of such Token whose span starts on selection
+                    var wantedNode = TryGetAcceptedNodeOrExtracted(rightNode, predicate, extractNode, syntaxFacts);
+                    if (wantedNode != default)
+                    {
+                        return wantedNode;
+                    }
+
+                    rightNode = rightNode?.Parent;
+                }
+                while (rightNode != null && rightNode.Span.Start == selection.Start);
+            }
+
+            // if selection inside tokenToRightOrIn -> no Token can be to Left (tokenToRightOrIn is left from selection)
+            if (tokenToRightOrIn != default && tokenToRightOrIn.Span.Start != selectionStripped.Start)
+            {
+                return default;
+            }
+
+            // Token to left
+            var tokenPreSelection = (tokenOnSelection.Span.End == selectionStripped.Start)
+                ? tokenOnSelection
+                : tokenOnSelection.GetPreviousToken();
+
+            var tokenToLeft = (tokenPreSelection.Span.End == selectionStripped.Start)
+                ? tokenPreSelection
+                : default;
+
+            if (tokenToLeft != default)
+            {
+                var leftNode = tokenToLeft.Parent;
+                do
+                {
+                    // consider either a Node that is a parent of touched Token (selection can be within) or ancestor Node of such Token whose span ends on selection
+                    var wantedNode = TryGetAcceptedNodeOrExtracted(leftNode, predicate, extractNode, syntaxFacts);
+                    if (wantedNode != default)
+                    {
+                        return wantedNode;
+                    }
+
+                    leftNode = leftNode?.Parent;
+                }
+                while (leftNode != null && leftNode.Span.End == selection.Start);
+            }
+
+            // nothing found
+            return default;
+
+            static SyntaxNode TryGetAcceptedNodeOrExtracted(SyntaxNode node, Predicate<SyntaxNode> predicate, Func<SyntaxNode, ISyntaxFactsService, SyntaxNode> extractNode, ISyntaxFactsService syntaxFacts)
+            {
+                if (node == default)
+                {
+                    return default;
+                }
+
+                if (predicate(node))
+                {
+                    return node;
+                }
+
+                var extrNode = extractNode(node, syntaxFacts);
+                if (extrNode != default && predicate(extrNode))
+                {
+                    return extrNode;
+                }
+
+                return default;
+            }
+        }
+
+        public virtual SyntaxNode DefaultNodeExtractor(SyntaxNode node, ISyntaxFactsService syntaxFacts)
+        {
+            // var a = b;
+            // -> b
```

Comment:
this warrants a better explanation.  i.e. the gut reason why we like this.  it can be as simple as:  "It's very reasonable for peopel to select an entire local-decl statement when wanting to just operate on the initializer section.  So we check for that, and consider such a yadda yadda".

Note: while you're here, shouldn't we do:

```c#
b;
-> b
```

and

```c#
a = b;  
-> b
```

?

i.e. a selection of an expr-statement should be considered a reasonable selection of the expr.  And the selection of an assignment should be considered a reasonable selection of the rhs?

Reply:
oops, i see you take care of assignment below.  How about expr-statement though?

Reply:
1) added explanation
2) the second one is covered, let's do the first one once we convert refactoring that would need it.

Reply:
Oh, I see the expr statement could actually be useful even for the already done refactorings. Yep, will fix.

PS: It seems that GitHub sometimes doesn't show new comments immediately.

Reply:
Actually  none of the refactorings we moved can be applicable for Expr-statement (LINQ can't be used in that way (CS0201) and I'd prefer having a testcase for that so let's wait for refactoring where it would be applicable.